File: Leetcode/binary-search/74-search-2D-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        lo = 0
        hi = m * n - 1

        while lo <= hi:
            mid = lo + (hi - lo) // 2
            r = mid // n
            c = mid % n
            if matrix[r][c] > target:
                hi = mid - 1
            elif matrix[r][c] < target:
                lo = mid + 1
            else:
                return True

        return False

File: Leetcode/binary-search/74-search-2D-matrix.cpp
class Solution {
  public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        int lo = 0;
        int hi = m * n - 1;

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int r = mid / n;
            int c = mid % n;
            if (target < matrix[r][c])
                hi = mid - 1;
            else if (target > matrix[r][c])
                lo = mid + 1;
            else
                return true;
        }

        return false;
    }
};
